ahnlich/typegen/src/tracers/mod.rs

```diff
@@ -0,0 +1,127 @@
+    let query_file = std::fs::File::create(file_path).expect("Failed to create query file");
+    let buffer = std::io::BufWriter::new(query_file);
+
+    serde_json::to_writer_pretty(buffer, &registry)
+        .expect("Query: Failed to write tracer registry into json file");
+}
+
+pub(crate) fn generate_language_definition(
+    language: Language,
+    input_dir: std::path::PathBuf,
+    output_dir: std::path::PathBuf,
+) {
+    let task = TypeGenTask::build(language, input_dir, output_dir);
+    task.generate_type_def_for_language()
+}
+
+struct TypeGenTask {
+    language: Language,
+    input_dir: std::path::PathBuf,
+    output_dir: std::path::PathBuf,
+}
+impl TypeGenTask {
+    fn build(
+        language: Language,
+        input_dir: std::path::PathBuf,
+        output_dir: std::path::PathBuf,
+    ) -> Self {
+        Self {
+            language,
+            input_dir,
+            output_dir,
+        }
+    }
+
+    fn generate_type_def_for_language(&self) {
+        let dir_entries = self
+            .input_dir
+            .as_path()
+            .read_dir()
+            .expect("Failed to read input dir")
+            .map(|entry| {
+                let file_path = entry.unwrap().path();
+                let file_stem = file_path.file_stem().unwrap();
+                let str_filename = file_stem.to_str().unwrap().to_string();
+                str_filename
+            });
+        for file_name in dir_entries.into_iter() {
+            let registry = load_type_into_registry(
+                self.input_dir
+                    .as_path()
+                    .join(format!("{}.json", &file_name)),
+            );
+            self.process_type_gen(&file_name, &registry);
+        }
+    }
+
+    fn process_type_gen(&self, file_name: &str, registry: &Registry) {
+        let config = serde_generate::CodeGeneratorConfig::new("".to_string())
+            .with_encodings(vec![serde_generate::Encoding::Bincode]);
+        match self.language {
+            Language::Python => {
+                let query_file =
+                    std::fs::File::create(self.output_dir.join(format!("{}.py",file_name)))
+                    .expect("Failed to create file");
+                let mut buffer: BufWriter<File> = std::io::BufWriter::new(query_file);
+
+                serde_generate::python3::CodeGenerator::new(&config)
+                    .output(&mut buffer, registry)
+            }
+            Language::Golang => {
+                let query_file =
+                    std::fs::File::create(self.output_dir.join(format!("{}.go",file_name)))
+                    .expect("Failed to create file");
+                let mut buffer: BufWriter<File> = std::io::BufWriter::new(query_file);
+                serde_generate::golang::CodeGenerator::new(&config)
+                    .output(&mut buffer, registry)
+            }
+            Language::Ocaml => {
+                let query_file =
```

Comment:
You can do this entire thing outside

If there is a  

```rust
struct OutputFile {
 output_file: std::path::PathBuf,
 language: Language,
}

impl OutputFile {
 fn generate(&self, config, registry) -> std::fs::File {
  let query_file = std::fs::File::create(&self.output_file).expect("Failed to create typegen output file");
  let mut buffer = BufWriter<File> = std::io::BufWriter::new(query_file);
  match self.language {
  Language::Python => serde_generate::python3::CodeGenerator::new(config).output(&mut buffer, registry),
 Language::Golang => serde_generate::go::CodeGenerator::new(config).output(&mut buffer, registry), 
 }
}

impl From<(&std::path::PathBuf, &Language)> for OutputFile {
 fn from((file_dir, file_name, language): (&std::path::PathBuf, &str, &Language)) -> OutputFile {
  let extension = match language {
  Language::Python => "py",
  Language::Go => "go",
  ...
  };
  OutputFile(file_dir.join(format!("{file_name}.go")))
 }
}
```

And then process typegen can look a bit simpler like

```rust
fn process_type_gen(file_dir, file_name) {
   if !file_dir.is_dir() {
   // maybe this would have already been checked at a higher layer
   // else check here and panic
  }
  let query_file: OutputFile = (file_dir, file_name, &self.language).into();
 query_file.generate(&config, &registry);
}
// or modify generate_type_def_for_language to do this directly
```